### PR TITLE
Automatic timeout from rlimits

### PR DESCRIPTION
**Pull request by gernst**
Leave some time at the end to write test suite

### CODE REVIEW

File: benchexec/tools/cmaesfuzz.py [hunk 1]
```diff
@@ -29,6 +29,16 @@ def executable(self, tool_locator):
         return tool_locator.find_executable("fuzzer")
 
     def cmdline(self, executable, options, task, rlimits):
+        # add a time limit if not given
+        # that is hopefully sufficient to write all tests
+        if "-t" not in options:
```
Comment (PhilippWendler) [inline benchexec/tools/cmaesfuzz.py:34]:
```suggestion
        if "-t" not in options and rlimits.cputime:
```
A time limit does not always exist.

File: benchexec/tools/cmaesfuzz.py [hunk 2]
```diff
@@ -29,6 +29,16 @@ def executable(self, tool_locator):
...
+            # at least three seconds + 1% of overall time
+            timeout = int(rlimits.cputime*0.99 - 5)
```
Comment (PhilippWendler) [inline benchexec/tools/cmaesfuzz.py:36]: Mismatch with comment?

File: benchexec/tools/cmaesfuzz.py [hunk 3]
```diff
@@ -29,6 +29,16 @@ def executable(self, tool_locator):
...
+            # but don't add negative timeout
+            if timeout > 0:
+                options = options + ["-t", timeout]
```
Comment (PhilippWendler) [inline benchexec/tools/cmaesfuzz.py:39]:
```suggestion
                options = options + ["-t", str(timeout)]
```

File: benchexec/tools/cmaesfuzz.py [hunk 4]
```diff
@@ -29,6 +29,16 @@ def executable(self, tool_locator):
...
+            else:
+                options = options + ["-t", rlimits.cputime]
```
Comment (PhilippWendler) [inline benchexec/tools/cmaesfuzz.py:41]:
```suggestion
                options = options + ["-t", str(rlimits.cputime)]
```

**Review by PhilippWendler — CHANGES_REQUESTED** [review]
Please test my suggestions. Didn't this crash in your tests before the `str()` calls are added? Would be weird if it didn't...

File: benchexec/tools/cmaesfuzz.py (hunk 2 above)
Comment (PhilippWendler) [reply benchexec/tools/cmaesfuzz.py:36]: Maybe you would also like to implement this offset within your tool? I.e., just passing the actual time limit to the tool, and the tool itself then deduces when it should stop?

Might be easier to update that way and more convenient for users who use the tool outside of BenchExec. The tool could also make a better estimate, i.e., depending on the size of the test suite... Or update the calculation if later versions are faster/slower.

Comment (gernst) [reply benchexec/tools/cmaesfuzz.py:36]: Maybe, but I'd rather not (the tool's time limit should be interpreted for finding test-cases, not writing them out)

Comment (PhilippWendler) [reply benchexec/tools/cmaesfuzz.py:36]: As you prefer.

But please make the comment consistent with the code and confirm that it works in tests now.

**Comment by dbeyer** [comment]
@gernst and @PhilippWendler There are unresolved conversations. Anything to add?

**Comment by PhilippWendler** [comment]
Yes, waiting for https://github.com/sosy-lab/benchexec/pull/675#discussion_r529677243 to be addressed.

**Comment by gernst** [comment]
Sure. I'll get to it soon

**Comment by gernst** [comment]
Ok this is fixed and tested now, hence ready to merge, thanks for the feedback.

**Review by PhilippWendler — APPROVED** [review]

**Pull request merged**